### PR TITLE
IGNITE-23599 Automatic reset for table partitions of HA zones

### PR DESCRIPTION
**Pull request by kgusakov**
https://issues.apache.org/jira/browse/IGNITE-23599

### CODE REVIEW

**Review by alievmirza — COMMENTED** [review]
I would expect at least 2-3 more tests: 
two majority lost in a row, correct behaviour of rebalances after majority lost (majority lost -> scale up, majority lost -> filter change)

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java [hunk 1]
```diff
@@ -248,29 +252,51 @@ public List<SystemView<?>> systemViews() {
...
+        return ongoingOperationsById;
+    }
+
     private CompletableFuture<Boolean> onHaZoneTopologyReduce(HaZoneTopologyUpdateEventParams params) {
         int zoneId = params.zoneId();
         long timestamp = params.timestamp();
 
         CatalogZoneDescriptor zoneDescriptor = catalogManager.zone(zoneId, timestamp);
         int catalogVersion = catalogManager.activeCatalogVersion(timestamp);
+        long revision = metaStorageManager.revisionByTimestampLocally(HybridTimestamp.hybridTimestamp(timestamp));
```
[the remark this answers is not in the document]
Comment (kgusakov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java]: Done. Redundant changes on metastorage removed also.

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java [hunk 2]
```diff
@@ -248,29 +252,51 @@ public List<SystemView<?>> systemViews() {
...
         List<CatalogTableDescriptor> tables = findTablesByZoneId(zoneId, catalogVersion, catalogManager);
         List<CompletableFuture<Void>> tablesResetFuts = new ArrayList<>();
         for (CatalogTableDescriptor table : tables) {
             Set<Integer> partitionsToReset = new HashSet<>();
             for (int partId = 0; partId < zoneDescriptor.partitions(); partId++) {
-                // TODO: https://issues.apache.org/jira/browse/IGNITE-23599 implement check for majority loss
-                partitionsToReset.add(partId);
+                TablePartitionId partitionId = new TablePartitionId(table.id(), partId);
+
+                if (stableAssignmentsWithOnlyAliveNodes(partitionId, revision).size() < zoneDescriptor.replicas() / 2 + 1) {
```
[the remark this answers is not in the document]
Comment (kgusakov) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/DisasterRecoveryManager.java]: Added

File: modules/table/src/integrationTest/java/org/apache/ignite/internal/table/distributed/disaster/ItHighAvailablePartitionsRecoveryTest.java [hunk 3]
```diff
@@ -202,7 +244,170 @@ private void checkRecoveryRequestOnlyOne(IgniteImpl node) {
...
+        }, 10_000));
+
+        assert revision.get() != 0;
+
+        return revision.get();
+    }
+
+    private void awaitForAllNodesTableGroupInitialization() throws InterruptedException {
+        assertTrue(waitForCondition(() -> {
+            AtomicLong numberOfInitializedReplicas = new AtomicLong(0);
```
[the remark this answers is not in the document]
Comment (kgusakov) [reply modules/table/src/integrationTest/java/org/apache/ignite/internal/table/distributed/disaster/ItHighAvailablePartitionsRecoveryTest.java]: Done

**Comment by kgusakov** [comment]
> I would expect at least 2-3 more tests: two majority lost in a row, correct behaviour of rebalances after majority lost (majority lost -> scale up, majority lost -> filter change)

I added the tests for scaleUp. But the test for filter change needs another number of nodes and it's configurations (attributes), I also prepared the infra for extending HA suites in this manner in the next PR https://github.com/apache/ignite-3/pull/4843/files#diff-0598f8c3fb6a622b56b139c6f0cdac49329b3963fe1c7057e21a0aed7bdfe0c2R54 - can I add the filter test in the next PR too. Just don't want to move this code in this PR and increase its size

**Review by alievmirza — APPROVED** [review]

**Pull request merged**